core/src/oidc_auth.rs

```diff
@@ -41,9 +42,38 @@ pub enum Error {
     NoIdpServerInfo,
     CsrfMismatch,
     HumanFlowUnsupported,
+    Timedout,
     Other(String),
 }
 
+impl From<Error> for mongodb::error::Error {
+    fn from(e: Error) -> Self {
+        mongodb::error::Error::custom(e)
+    }
+}
+
+pub async fn oidc_call_back(params: CallbackContext) -> mongodb::error::Result<IdpServerResponse> {
+    let sleep_duration = params
+        .timeout_seconds
+        // turn the supplied timeout Instant into a Duration from now
+        .map(|x| x - Instant::now())
+        .unwrap_or(DEFAULT_SLEEP_DURATION);
+    let timeout = time::sleep(sleep_duration);
+    tokio::pin!(timeout);
+
+    if params.refresh_token.is_some() {
```

Comment:
I get the idea behind this now that I looked at the doc, but couldn't we have a simpler and more straightforward way of having a timeout on a job?

It looks like the [`time` feature from Tokio](https://docs.rs/tokio/latest/tokio/time/index.html) would work well here.
For example 
```
let res = timeout(Duration::from_secs(1), do_refresh(params)).await;

if res.is_err() {
    println!("operation timed out");
}
```

Reply:
I wrote it this way because this is how you'd write it in go :joy:. Had no idea this timeout function worked this way.